safenode/src/network/mod.rs

```diff
@@ -296,50 +300,31 @@ impl Network {
         self.get_closest_peers(xor_name, false).await
     }
 
-    /// Returns the closest peers to the given `XorName`, sorted by their distance to the xor_name.
-    /// If `client` is false, then include `self` among the `closest_peers`
-    async fn get_closest_peers(&self, xor_name: XorName, client: bool) -> Result<Vec<PeerId>> {
-        let (sender, receiver) = oneshot::channel();
-        self.send_swarm_cmd(SwarmCmd::GetClosestPeers { xor_name, sender })
-            .await?;
-        let (our_id, k_bucket_peers) = receiver.await?;
-
-        // Count self in if among the CLOSE_GROUP_SIZE closest and sort the result
-        let mut closest_peers: Vec<_> = k_bucket_peers.into_iter().collect();
-        if !client {
-            closest_peers.push(our_id);
-        }
-        let target = KBucketKey::new(xor_name.0.to_vec());
-        closest_peers.sort_by(|a, b| {
-            let a = KBucketKey::new(a.to_bytes());
-            let b = KBucketKey::new(b.to_bytes());
-            target.distance(&a).cmp(&target.distance(&b))
-        });
-        let closest_peers: Vec<PeerId> = closest_peers
-            .iter()
-            .take(CLOSE_GROUP_SIZE)
-            .cloned()
-            .collect();
-
-        if CLOSE_GROUP_SIZE > closest_peers.len() {
-            warn!("Not enough peers in the k-bucket to satisfy the request");
-            return Err(Error::NotEnoughPeers);
-        }
-
-        trace!(
-            "Got the {} closest_peers to the given XorName-{xor_name}, nodes: {closest_peers:?}",
-            closest_peers.len()
+    /// Send `Request` to the closest peers. `Self` might be present among the recipients.
+    pub async fn node_send_to_closest(&self, request: &Request) -> Result<Vec<Result<Response>>> {
+        info!(
+            "Sending {request:?} with dst {:?} to the closest peers.",
+            request.dst().name()
         );
+        // todo: if `self` is present among the closest peers, the request should be routed to self?
+        let closest_peers = self.node_get_closest_peers(*request.dst().name()).await?;
```

Comment:
> Just to confirm that we can send any number of msgs to ourselves and it will just flow through without any deadlock or funny loops or such.

The msg to `self` indeed causes an issue, as confirmed by a test. I have updated the PR with a fix for it.